src/cmt_filter.c

```diff
@@ -0,0 +1,316 @@
+            continue;
+        }
+
+        ret = cmt_cat_counter(dst, counter);
+        if (ret == -1) {
+            return -1;
+        }
+    }
+
+    /* Gauges */
+    cfl_list_foreach(head, &src->gauges) {
+        gauge = cfl_list_entry(head, struct cmt_gauge, _head);
+
+        if (compare_label_keys(gauge->map, label_key, compare_ctx, compare, flags) == CMT_FALSE) {
+            continue;
+        }
+
+        ret = cmt_cat_gauge(dst, gauge);
+        if (ret == -1) {
+            return -1;
+        }
+    }
+
+    /* Untyped */
+    cfl_list_foreach(head, &src->untypeds) {
+        untyped = cfl_list_entry(head, struct cmt_untyped, _head);
+
+        if (compare_label_keys(untyped->map, label_key, compare_ctx, compare, flags) == CMT_FALSE) {
+            continue;
+        }
+
+        ret = cmt_cat_untyped(dst, untyped);
+        if (ret == -1) {
+            return -1;
+        }
+    }
+
+    /* Histogram */
+    cfl_list_foreach(head, &src->histograms) {
+        histogram = cfl_list_entry(head, struct cmt_histogram, _head);
+
+        if (compare_label_keys(histogram->map, label_key, compare_ctx, compare, flags) == CMT_FALSE) {
+            continue;
+        }
+
+        ret = cmt_cat_histogram(dst, histogram);
+        if (ret == -1) {
+            return -1;
+        }
+    }
+
+    /* Summary */
+    cfl_list_foreach(head, &src->summaries) {
+        summary = cfl_list_entry(head, struct cmt_summary, _head);
+
+        if (compare_label_keys(summary->map, label_key, compare_ctx, compare, flags) == CMT_FALSE) {
+            continue;
+        }
+
+        ret = cmt_cat_summary(dst, summary);
+        if (ret == -1) {
+            return -1;
+        }
+    }
+
+    return CMT_FILTER_SUCCESS;
+}
+
+static int compare_fqname(struct cmt_opts *src, const char *fqname,
+                          void *compare_ctx, int (*compare)(void *compare_ctx, const char *str, size_t slen),
+                          int flags)
+{
+    /* compare fqname for prefix */
+    if (flags & CMT_FILTER_PREFIX) {
+        if (strncmp(src->fqname, fqname, strlen(fqname)) == 0) {
+            return (flags & CMT_FILTER_EXCLUDE) ? CMT_FALSE : CMT_TRUE;
+        }
+
+        return (flags & CMT_FILTER_EXCLUDE) ? CMT_TRUE : CMT_FALSE;
+    }
+
+    /* compare fqname for substring */
+    if (flags & CMT_FILTER_SUBSTRING) {
+        if (strstr(src->fqname, fqname) != NULL) {
+            return (flags & CMT_FILTER_EXCLUDE) ? CMT_FALSE : CMT_TRUE;
+        }
+
+        return (flags & CMT_FILTER_EXCLUDE) ? CMT_TRUE : CMT_FALSE;
+    }
+
+    /* Compare with an external context (e.g. Onigmo).
+     * flb_regex_match should take three arguments that are
+     * flb_regex context, string and its string length.
+     * The length of string is changed by the callback and not determined by fqname.
+     */
+    if (compare_ctx != NULL && compare != NULL) {
+        return compare(compare_ctx, src->fqname, strlen(src->fqname));
```

Comment:
is fqname an sds ?

Reply:
It not always sds. Because compare is an external function and it depends on registered function.

Reply:
Once I rethought this, I realized this could be calculated before it's needed. So, prefix matching case and using an external function case should be able to calculate size of fqname. So, the previous my comment is wrong. Thanks for your suggestion.